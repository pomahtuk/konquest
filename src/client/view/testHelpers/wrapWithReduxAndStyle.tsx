import React, { ReactElement } from "react";
import { Provider } from "react-redux";
import storeCreator from "../../stores/game.store";
import ThemeProvider from "../themes/ThemeProvider";
import baseTheme from "../themes/base.theme";
const store = storeCreator();

export const CurrentStore = store;

const wrapWithReduxAndStyle = (element: ReactElement): ReactElement => {
  return (
    <Provider store={store}>
      <ThemeProvider theme={baseTheme}>{element}</ThemeProvider>
    </Provider>
  );
};

export default wrapWithReduxAndStyle;
