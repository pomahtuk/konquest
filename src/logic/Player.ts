export default class Player {
  public id: number;
  public screenName: string;

  public constructor(name: string) {
    // TODO: player actually have to be stored somewhere
    this.id = Math.floor(Math.random() * 1000);
    this.screenName = name;
  }
}

export interface PlayerMap {
  [key: string]: Player;
}
