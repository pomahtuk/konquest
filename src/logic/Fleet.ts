import Player from "./Player";

export interface FleetParams {
  amount: number;
  killPercent?: number;
  owner: Player;
  destination: string;
}

export default class Fleet {
  public amount: number;
  public killPercent: number;
  public owner: Player;
  public destination: string;

  public constructor({ owner, killPercent = 0.5, amount, destination }: FleetParams) {
    this.owner = owner;
    this.killPercent = killPercent;
    this.amount = amount;
    this.destination = destination;
  }
}
