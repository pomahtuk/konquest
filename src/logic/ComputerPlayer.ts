import Player, { PlayerTurnOrder } from "./Player";
import { PlanetMap } from "./Planet";
import Fleet from "./Fleet";

export enum ComputerPlayerType {
  EASY = "easy",
  NORMAL = "normal",
  HARD = "hard"
}

class ComputerPlayer extends Player {
  public computerType: ComputerPlayerType;

  constructor(name: string, type: ComputerPlayerType) {
    super(name);
    this.isComputer = true;
    this.computerType = type;
  }

  public shouldSkipPlanet(planetName: string, orders: PlayerTurnOrder[], fleets: Fleet[]): boolean {
    for (const order of orders) {
      if (order.destination === planetName) {
        return true;
      }
    }

    for (const fleet of fleets) {
      if (fleet.destination === planetName) {
        return true;
      }
    }

    return false;
  }

  public takeTurn(planets: PlanetMap, fleets: Fleet[]): PlayerTurnOrder[] {
    return [];
  }
}

export default ComputerPlayer;
