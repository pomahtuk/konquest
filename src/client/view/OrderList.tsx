import React, { ReactElement, useContext } from "react";
import { PlayerTurnOrder } from "../../logic/Player";
import { ThemeContext } from "./themes/ThemeProvider";
import { css } from "emotion";

export interface OrderListProps {
  orders: PlayerTurnOrder[];
  removeOrder: (index: number) => void;
}

const OrderList = ({ orders, removeOrder }: OrderListProps): ReactElement | null => {
  const theme = useContext(ThemeContext);

  if (!orders || orders.length === 0) {
    return null;
  }

  const spanStyle = css`
    cursor: pointer;
    margin: 0 0 0 ${theme.units.medium};
    font-size: ${theme.fontSizes.smallest};
  `;

  return (
    <ul>
      {orders.map((order, index) => (
        <li key={`${order.amount}${order.origin}${order.destination}`}>
          <span>
            Send fleet({order.amount}) from {order.origin} to {order.destination}
          </span>
          <span className={spanStyle} onClick={() => removeOrder(index)} data-testid="remove">
            ‚ùå
          </span>
        </li>
      ))}
    </ul>
  );
};

export default OrderList;
