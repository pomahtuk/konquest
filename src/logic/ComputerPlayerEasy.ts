import ComputerPlayer, { ComputerPlayerType } from "./ComputerPlayer";

class ComputerPlayerEasy extends ComputerPlayer {
  protected minimumShips = 20;
  protected shipCountFactor = 2;

  constructor(name: string) {
    super(name, ComputerPlayerType.EASY);
  }
}

export default ComputerPlayerEasy;
