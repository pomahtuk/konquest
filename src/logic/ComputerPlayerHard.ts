import ComputerPlayer, { ComputerPlayerType } from "./ComputerPlayer";

class ComputerPlayerHard extends ComputerPlayer {
  protected minimumShips = 30;
  protected shipCountFactor = 3;

  constructor(name: string) {
    super(name, ComputerPlayerType.HARD);
  }
}

export default ComputerPlayerHard;
