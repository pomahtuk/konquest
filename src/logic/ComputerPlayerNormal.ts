import ComputerPlayer, { ComputerPlayerType } from "./ComputerPlayer";

class ComputerPlayerNormal extends ComputerPlayer {
  protected minimumShips = 10;
  protected shipCountFactor = 2;

  constructor(name: string) {
    super(name, ComputerPlayerType.NORMAL);
  }
}

export default ComputerPlayerNormal;
