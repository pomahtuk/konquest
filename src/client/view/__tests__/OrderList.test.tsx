import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { matchers } from "jest-emotion";

expect.extend(matchers);

import wrapWithReduxAndStyle from "../testHelpers/wrapWithReduxAndStyle";
import OrderList from "../OrderList";
import { PlayerTurnOrder } from "../../../logic/Player";

const removeOrder = jest.fn();

describe("<OrderList />", (): void => {
  it("renders <OrderList /> component without any orders", (): void => {
    const { container } = render(wrapWithReduxAndStyle(<OrderList orders={[]} removeOrder={removeOrder} />));
    expect(container).toMatchSnapshot();
  });

  it("renders <OrderList /> component with order", (): void => {
    const order: PlayerTurnOrder = { origin: "A", destination: "B", amount: 10 };
    const { container } = render(wrapWithReduxAndStyle(<OrderList orders={[order]} removeOrder={removeOrder} />));
    expect(container).toMatchSnapshot();
  });

  it("<OrderList /> calling remove callback", (): void => {
    const order: PlayerTurnOrder = { origin: "A", destination: "B", amount: 10 };
    const { getByTestId } = render(wrapWithReduxAndStyle(<OrderList orders={[order]} removeOrder={removeOrder} />));
    const removeButton = getByTestId("remove");
    fireEvent.click(removeButton);
    expect(removeOrder).toHaveBeenCalledTimes(1);
  });
});
