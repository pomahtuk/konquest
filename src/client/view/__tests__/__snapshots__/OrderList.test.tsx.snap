// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OrderList /> renders <OrderList /> component with order 1`] = `
.emotion-0 {
  cursor: pointer;
  margin: 0 0 0 8px;
  font-size: 10px;
}

<div>
  <ul>
    <li>
      <span>
        Send fleet(
        10
        ) from 
        A
         to 
        B
      </span>
      <span
        class="emotion-0"
        data-testid="remove"
      >
        ‚ùå
      </span>
    </li>
  </ul>
</div>
`;

exports[`<OrderList /> renders <OrderList /> component without any orders 1`] = `<div />`;
